lib/eigensolve_quda.cpp

```diff
@@ -0,0 +1,948 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <math.h>
+#include <iostream>
+#include <vector>
+#include <algorithm>
+
+#include <quda_internal.h>
+#include <eigensolve_quda.h>
+#include <qio_field.h>
+#include <color_spinor_field.h>
+#include <blas_quda.h>
+#include <util_quda.h>
+#include <sys/time.h>
+
+#include <Eigen/Eigenvalues>
+#include <Eigen/Dense>
+
+bool flags = true;
+
+namespace quda
+{
+
+  using namespace Eigen;
+
+  // Eigensolver class
+  //-----------------------------------------------------------------------------
+  EigenSolver::EigenSolver(QudaEigParam *eig_param, TimeProfile &profile)
+    : eig_param(eig_param), profile(profile), tmp1(nullptr), tmp2(nullptr)
+  {
+    // Timings for components of the eigensolver
+    time_ = 0.0;
+    time_e = 0.0;   // time in Eigen
+    time_mv = 0.0;  // time in matVec
+    time_mb = 0.0;  // time in multiblas
+    time_svd = 0.0; // time to compute SVD
+
+    // Problem parameters
+    nEv = eig_param->nEv;
+    nKr = eig_param->nKr;
+    nConv = eig_param->nConv;
+    tol = eig_param->tol;
+    reverse = false;
+
+    // Algorithm variables
+    converged = false;
+    restart_iter = 0;
+    max_restarts = eig_param->max_restarts;
+    check_interval = eig_param->check_interval;
+    iter = 0;
+    iter_converged = 0;
+    iter_locked = 0;
+    iter_keep = 0;
+    num_converged = 0;
+    num_locked = 0;
+    num_keep = 0;
+
+    // Sanity checks
+    if (nKr <= nEv) errorQuda("nKr=%d is less than or equal to nEv=%d\n", nKr, nEv);
+    if (nEv < nConv) errorQuda("nConv=%d is greater than nEv=%d\n", nConv, nEv);
+    if (nEv == 0) errorQuda("nEv=0 passed to Eigensolver\n");
+    if (nKr == 0) errorQuda("nKr=0 passed to Eigensolver\n");
+    if (nConv == 0) errorQuda("nConv=0 passed to Eigensolver\n");
+
+    residua = (double*)safe_malloc(nKr * sizeof(double));
+    for (int i = 0; i < nKr; i++) { residua[nKr] = 0.0; }
+
+    // Quda MultiBLAS freindly array
+    Qmat = (Complex*)safe_malloc(nEv * nKr * sizeof(Complex));
+
+    // Part of the spectrum to be computed.
+    spectrum = strdup("SR"); // Initialised to stop the compiler warning.
+
+    if (eig_param->use_poly_acc) {
```

Comment:
In the case of polynomial acceleration: do we want to make anything that's not `SR` throw an error? (except maybe `LR`, which may be worth a warning instead---it's highly suboptimal but it's not really wrong if I remember correctly). 

There's a big asterisk here in that iirc Chebyshev acceleration can be proven to work for spectra that aren't purely real as long as they lie in a well-behaved ellipse in the right half plane (i.e., I believe you can get it to work on a Wilson-clover spectrum with enough experimenting), so that'd motivate just letting this be for now...

Reply:
Ok, leaving this alone for the moment.